### PR TITLE
feat: segregate connect/auth/refresh/enable device duties

### PR DESCRIPTION
**Pull request by chemelli74**
## Summary by CodeRabbit

- **New Features**
	- Enhanced device connection reliability with a retry mechanism for socket connections.
	- Introduced a new method for streamlined authentication and status refresh.
	- Improved logging for device state changes and authentication processes.
	- Added a new exception class for better error handling related to connection failures.

- **Bug Fixes**
	- Addressed potential transient connection failures through retry logic.

- **Tests**
	- Refactored tests for improved clarity and maintainability using parameterization.

### CODE REVIEW

File: midealocal/device.py [hunk 1]
```diff
@@ -510,10 +512,9 @@ def _check_heartbeat(self, now: float) -> None:
     def run(self) -> None:
         """Run loop."""
         while self._is_run:
-            while self._socket is None:
-                if self.connect(refresh_status=True) is False:
-                    self.close_socket()
-                    time.sleep(5)
+            if not self._socket or not self.connect():
```
Comment (rokam) [inline midealocal/device.py:515]: This looks a little bit off. It'll call connect if socket is not none and a new socket will be open.
I didn't put my thought in this but this should be something like this:
```suggestion
            if self._socket is None and not self.connect():
```

**Review by rokam — CHANGES_REQUESTED** [review]
I saw just a possible issue.

File: midealocal/device.py (hunk 1 above)
Comment (chemelli74) [reply midealocal/device.py:515]: `if not self._socket` gives the same result as `if self._socket is None`

Comment (rokam) [reply midealocal/device.py:515]: Anyway, it should use ` and` instead of `or`

Comment (chemelli74) [reply midealocal/device.py:515]: not convinced at all, but I'll change it if you think it should be changed.

Comment (rokam) [reply midealocal/device.py:515]: if socket is not none it'll go to connect that will replace the socket, leaving an open socket.

Comment (chemelli74) [reply midealocal/device.py:515]: With suggested change, mypy complains:

image

Comment (chemelli74) [reply midealocal/device.py:515]: > if socket is not none it'll go to connect that will replace the socket, leaving an open socket.

For this we can just add a : 

```python
finally:
                self._socket.close()
```

in the connect() try block

Comment (rokam) [reply midealocal/device.py:518]: You didn't get it.

If you use `or`:
| socket is None | connect return | result |
|--------|--------|--------|
| True | Won't be run | won't raise socketexception |
| False | False | There is an active socket and another one is instantiated inside connect function, then socketexception is raised |
| False | True | There is an active socket and another one is instantiated inside connect function, no socketexception is raised |

Comment (chemelli74) [reply midealocal/device.py]: Reworked to improve socket management, even if I think it's still not optimal.
Enough to proceed with this PR.

**Review by rokam — APPROVED** [review]
LGTM

**Pull request merged**